packages/firestore/src/api/user_data_converter.ts

```diff
@@ -297,19 +299,54 @@ export class UserDataConverter {
   }
 
   /** Parse document data from a set() call with '{merge:true}'. */
-  parseMergeData(methodName: string, input: AnyJs): ParsedSetData {
+  parseMergeData(
+    methodName: string,
+    input: AnyJs,
+    fieldPaths?: Array<string | firestore.FieldPath>
+  ): ParsedSetData {
     const context = new ParseContext(
       UserDataSource.MergeSet,
       methodName,
       FieldPath.EMPTY_PATH
     );
     validatePlainObject('Data must be an object, but it was:', context, input);
 
-    const updateData = this.parseData(input, context);
-    const fieldMask = new FieldMask(context.fieldMask);
+    const updateData = this.parseData(input, context) as ObjectValue;
+    let fieldMask: FieldMask | null = null;
+
+    if (!fieldPaths) {
+      fieldMask = new FieldMask(context.fieldMask);
+    } else {
+      const validatedFieldPaths = [];
+
+      for (const stringOrFieldPath of fieldPaths) {
+        let fieldPath;
+
+        if (stringOrFieldPath instanceof ExternalFieldPath) {
```

Comment:
Note that there are 3 different FieldPath in our API: The internal representation, the external interface (firestore.FieldPath) and the backing type for the external interface (ExternalFieldPath). By default, a firestore.FieldPath is not compatible with an ExternalFieldPath as it has properties missing.